torch_xla/csrc/runtime/stablehlo_composite_helper.cc

```diff
@@ -0,0 +1,465 @@
+#include "torch_xla/csrc/runtime/stablehlo_composite_helper.h"
+
+#include <limits>
+#include <string>
+#include <tuple>
+#include <unordered_map>
+
+#include "absl/log/log.h"
+#include "absl/strings/str_cat.h"
+#include "mlir/IR/IRMapping.h"
+#include "mlir/Support/LogicalResult.h"
+#include "single_include/nlohmann/json.hpp"
+#include "stablehlo/dialect/StablehloOps.h"
+
+namespace torch_xla {
+namespace runtime {
+
+namespace {
+
+using nlohmann::json;
+
+static bool IsXlaMarkTensorOp(mlir::Operation* op) {
```

Comment:
Something we've found useful for CustomCall operations is using an adaptor. This does not have to happen now, just wanted to make you aware of this since it makes verification / accessors much easier:
https://github.com/tensorflow/mlir-hlo/blob/753fda7c2a79732eaa1f6eb076b1ac90c72441de/stablehlo/stablehlo/experimental/dialect/StablehloOps.h#L200

Then you have getters which take a custom call and return an optional adaptor if the CC is well-formed:
https://github.com/tensorflow/mlir-hlo/blob/753fda7c2a79732eaa1f6eb076b1ac90c72441de/stablehlo/stablehlo/experimental/dialect/StablehloOps.h#L224

Then you can write methods that take in this MarkTensorOp, and you don't need to verify the custom call everywhere

Reply:
Sounds good, I like this pattern. I created a separate bug to track the progress of this refactor.